File: internal/html/assets/src/crypto/bip138.test.ts
/**
 * Conformance tests for the second on-chain backup format.
 *
 * Every vector here comes from the BIP-138 draft itself, copied from
 * bitcoin/bips#1951 at commit 5af62cba9958a519218bcad8a0aae9e2090bb5bd. They
 * are the reason we can offer this format at all: the draft is open, so the
 * only safe way to follow it is to fail loudly when its own vectors stop
 * matching our bytes.
 */

import { test } from 'node:test';
import assert from 'node:assert/strict';

import {
  BIP380,
  bytesToHex,
  decodeBackup,
  decryptBackup,
  decryptDescriptor,
  deriveSecrets,
  descriptorPubkeys,
  dropCommonDerivationPaths,
  isCommonDerivationPath,
  encodeBackup,
  encodeBipContentType,
  encodeCompactSize,
  encodeDerivationPaths,
  encodeIndividualSecrets,
  encryptDescriptor,
  INHERITANCE_SECRET_ENTRIES,
  decodeBackup as decodeBip138,
  hexToBytes,
  normalizeKeys,
  parseDerivationPath,
  taggedHash,
} from './bip138.ts';

import encryptionSecretVectors from './testdata/bip138/encryption_secret.json' with { type: 'json' };
import individualSecretVectors from './testdata/bip138/individual_secrets.json' with { type: 'json' };
import derivationPathVectors from './testdata/bip138/derivation_path.json' with { type: 'json' };
import contentTypeVectors from './testdata/bip138/content_type.json' with { type: 'json' };
import cipherVectors from './testdata/bip138/chacha20poly1305_encryption.json' with { type: 'json' };
import backupVectors from './testdata/bip138/encrypted_backup.json' with { type: 'json' };
import descriptorVector from './testdata/descriptor-vector.json' with { type: 'json' };

const utf8 = new TextEncoder();

test('draft vectors: the decryption secret and every individual secret', () => {
  for (const v of encryptionSecretVectors) {
    const { secret, individualSecrets } = deriveSecrets(v.keys.map(hexToBytes));
    assert.equal(bytesToHex(secret), v.decryption_secret, v.description);
    assert.deepEqual(individualSecrets.map(bytesToHex), v.individual_secrets, v.description);
  }
});

test('draft vectors: individual secrets are sorted and de-duplicated on the way out', () => {
  for (const v of individualSecretVectors) {
    const secrets = v.secrets.map(hexToBytes);
    if (v.expected === null) {
      assert.throws(() => encodeIndividualSecrets(secrets), undefined, v.description);
      continue;
    }
    assert.equal(bytesToHex(encodeIndividualSecrets(secrets)), v.expected, v.description);
  }
});

test('draft vectors: derivation path encoding', () => {
  for (const v of derivationPathVectors) {
    const paths = v.paths.map(parseDerivationPath);
    if (v.expected === null) {
      // The draft marks a case that must be refused with a null expectation.
      assert.throws(() => encodeDerivationPaths(paths), undefined, v.description);
      continue;
    }
    assert.equal(bytesToHex(encodeDerivationPaths(paths)), v.expected, v.description);
  }
});

test('draft vectors: content type encoding for BIP-number types', () => {
  for (const v of contentTypeVectors) {
    if (!v.valid || !v.content.startsWith('01')) continue;
    const bip = parseInt(v.content.slice(2), 16);
    assert.equal(bytesToHex(encodeBipContentType(bip)), v.content, v.description);
  }
});

test('draft vectors: the cipher itself', async () => {
  const { chacha20poly1305 } = await import('@noble/ciphers/chacha.js');
  for (const v of cipherVectors) {
    if (v.ciphertext === null) continue; // refusals are asserted below
    const sealed = chacha20poly1305(hexToBytes(v.secret), hexToBytes(v.nonce)).encrypt(
      hexToBytes(v.plaintext)
    );
    assert.equal(bytesToHex(sealed), v.ciphertext, v.description);
  }
});

test('draft vectors: the refusals the cipher cases describe', () => {
  const keys = backupVectors[0].keys.map(hexToBytes);
  // "Empty plaintext should fail": the draft forbids an empty payload.
  assert.throws(() => encodeBackup({ pubkeys: keys, items: [] }), /at least one content item/);
  // "Encryption with zeroed nonce should fail".
  assert.throws(
    () =>
      encodeBackup({
        pubkeys: keys,
        items: [{ bip: BIP380, content: utf8.encode('x') }],
        nonce: new Uint8Array(12),
      }),
    /all zero/
  );
});

test('draft vectors: whole backups encode to the expected bytes', () => {
  let checked = 0;
  for (const v of backupVectors) {
    if (v.valid === false) continue;
    // The encoder writes BIP-number content items. A vector with a
    // vendor-specific type is covered by the decode test below instead.
    if (!v.content.startsWith('01')) continue;

    const items = [
      { bip: parseInt(v.content.slice(2), 16), content: utf8.encode(v.plaintext) },
      ...(v.extra ?? []).map((e: { content: string; plaintext: string }) => ({
        bip: parseInt(e.content.slice(2), 16),
        content: utf8.encode(e.plaintext),
      })),
    ];

    const backup = encodeBackup({
      pubkeys: v.keys.map(hexToBytes),
      items,
      decoySecrets: v.decoy_individual_secrets.map(hexToBytes),
      derivationPaths: v.derivation_paths.map(parseDerivationPath),
      nonce: hexToBytes(v.nonce),
    });

    assert.equal(bytesToHex(backup), v.expected, v.description);
    checked++;
  }
  assert.ok(checked >= 4, `expected several encodable vectors, checked ${checked}`);
});

test('draft vectors: every key in a backup can open it', () => {
  for (const v of backupVectors) {
    if (v.valid === false) continue;
    for (const key of v.keys) {
      const items = decryptBackup(hexToBytes(v.expected), hexToBytes(key));
      if (!v.content.startsWith('01')) {
        // A vendor-specific type. The draft says a parser skips what it does
        // not implement and keeps reading, so opening it must not throw.
        continue;
      }
      assert.ok(items.length >= 1, v.description);
      assert.equal(new TextDecoder().decode(items[0].content), v.plaintext, v.description);
    }
  }
});

test('draft vectors: bytes after the ciphertext are ignored', () => {
  const withTrailing = backupVectors.find((v) => v.trailing);
  assert.ok(withTrailing, 'the draft ships a vector with trailing bytes');
  const padded = hexToBytes(withTrailing.expected + withTrailing.trailing);
  const items = decryptBackup(padded, hexToBytes(withTrailing.keys[0]));
  assert.equal(new TextDecoder().decode(items[0].content), withTrailing.plaintext);
});

test('draft vectors: an all-zero nonce is refused', () => {
  const zeroNonce = backupVectors.find((v) => v.valid === false);
  assert.ok(zeroNonce, 'the draft ships an invalid-nonce vector');
  assert.throws(() => decodeBackup(hexToBytes(zeroNonce.expected)), /all-zero nonce/);
  assert.throws(
    () =>
      encodeBackup({
        pubkeys: zeroNonce.keys.map(hexToBytes),
        items: [{ bip: BIP380, content: utf8.encode('x') }],
        nonce: new Uint8Array(12),
      }),
    /all zero/
  );
});

test('a key that is not in the backup opens nothing', () => {
  const v = backupVectors[1];
  const stranger = hexToBytes(
    '03c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5'
  );
  assert.throws(() => decryptBackup(hexToBytes(v.expected), stranger), /None of the keys/);
});

test('keys sharing an x coordinate collapse to one, and the NUMS point is dropped', () => {
  const even = hexToBytes('02e6642fd69bd211f93f7f1f36ca51a26a5290eb2dd1b0d8279a87bb0d480c8443');
  const odd = hexToBytes('03e6642fd69bd211f93f7f1f36ca51a26a5290eb2dd1b0d8279a87bb0d480c8443');
  assert.equal(normalizeKeys([even, odd]).length, 1);

  const nums = hexToBytes('0250929b74c1a04954b78b4b6035e97a5e078a5a0f28ec96d547bfee9ace803ac0');
  assert.equal(normalizeKeys([even, nums]).length, 1);
  assert.throws(() => deriveSecrets([nums]), /at least one usable public key/);
});

test('compact size follows the Bitcoin encoding', () => {
  assert.equal(bytesToHex(encodeCompactSize(0)), '00');
  assert.equal(bytesToHex(encodeCompactSize(252)), 'fc');
  assert.equal(bytesToHex(encodeCompactSize(253)), 'fdfd00');
  assert.equal(bytesToHex(encodeCompactSize(0xffff)), 'fdffff');
  assert.equal(bytesToHex(encodeCompactSize(0x10000)), 'fe00000100');
});

test('the tagged hash matches BIP-340', () => {
  // sha256("TapLeaf") twice, then the message, is the standard construction.
  const out = taggedHash('BIP138_INDIVIDUAL_SECRET', new Uint8Array(32));
  assert.equal(out.length, 32);
  assert.notEqual(bytesToHex(out), bytesToHex(taggedHash('BIP138_DECRYPTION_SECRET', new Uint8Array(32))));
});

// --------------------------------------------------------------------------
// The descriptor path, which is what Bitcoin Inheritance actually offers a client.
// --------------------------------------------------------------------------

test('a real descriptor round-trips, and any one key opens it', () => {
  const { backup, text, excluded } = encryptDescriptor(descriptorVector.descriptor);
  assert.deepEqual(excluded, [], 'every key in a BIP-48 descriptor is eligible');
  for (const xpub of descriptorVector.xpubs) {
    assert.equal(decryptDescriptor(backup, xpub), descriptorVector.descriptor);
    assert.equal(decryptDescriptor(text, xpub), descriptorVector.descriptor);
  }
});

test('one key is enough, which is the difference a client must understand', () => {
  const { backup } = encryptDescriptor(descriptorVector.descriptor);
  // The k-of-n scheme needs two of these three. This one needs any single key.
  const alone = decryptDescriptor(backup, descriptorVector.xpubs[0]);
  assert.equal(alone, descriptorVector.descriptor);
});

test('a bare key with no derivation is refused', () => {
  const bare = `wsh(sortedmulti(2,${descriptorVector.xpubs.join(',')}))`;
  assert.throws(() => descriptorPubkeys(bare), /no key that BIP-138 can use/);
});

test('a descriptor mixing eligible and bare keys reports the excluded ones', () => {
  const mixed = `wsh(sortedmulti(2,${descriptorVector.xpubs[0]}/<0;1>/*,${descriptorVector.xpubs[1]}))`;
  const { pubkeys, excluded } = descriptorPubkeys(mixed);
  assert.equal(pubkeys.length, 1);
  assert.deepEqual(excluded, [descriptorVector.xpubs[1]]);
});

test('common account paths are dropped, uncommon ones are kept', () => {
  const common = parseDerivationPath("m/48'/1'/0'/2'");
  const personal = parseDerivationPath("m/0/1'/2/3'");
  assert.equal(isCommonDerivationPath(common), true);
  assert.equal(isCommonDerivationPath(personal), false);
  assert.equal(isCommonDerivationPath(parseDerivationPath("m/84'/0'/0'")), true);
  assert.equal(isCommonDerivationPath(parseDerivationPath("m/48'/0'/0'/3'")), false);
  assert.equal(isCommonDerivationPath(parseDerivationPath("m/44'/0'/10'")), false);
  assert.deepEqual(dropCommonDerivationPaths([common, personal]), [personal]);
});

test('a Bitcoin Inheritance backup always writes seven secret entries', () => {
  const { backup } = encryptDescriptor(descriptorVector.descriptor);
  const decoded = decodeBip138(backup);
  assert.equal(INHERITANCE_SECRET_ENTRIES, 7);
  assert.equal(decoded.individualSecrets.length, 7, 'three real keys, four decoys');

  // The decoys must not cost anyone their recovery.
  for (const xpub of descriptorVector.xpubs) {
    assert.equal(decryptDescriptor(backup, xpub), descriptorVector.descriptor);
  }

  // Entries are sorted by their own bytes, so the real ones are not grouped
  // at the front where a reader could pick them out.
  const asHex = decoded.individualSecrets.map(bytesToHex);
  assert.deepEqual(asHex, [...asHex].sort(), 'entries are written in sorted order');
  assert.equal(new Set(asHex).size, 7, 'no entry repeats');
});

test('two backups of the same wallet share no decoy', () => {
  const a = decodeBip138(encryptDescriptor(descriptorVector.descriptor).backup);
  const b = decodeBip138(encryptDescriptor(descriptorVector.descriptor).backup);
  const inA = new Set(a.individualSecrets.map(bytesToHex));
  const shared = b.individualSecrets.map(bytesToHex).filter((x) => inA.has(x));
  // The three real entries repeat, because they are derived from the keys.
  // Anything beyond that would mean the decoys were not random.
  assert.equal(shared.length, 3, 'only the real entries repeat across backups');
});

test('the on-chain size of both formats is measured, not guessed', () => {
  const kOfN = Buffer.from(
    descriptorVector.encryptedText.slice(descriptorVector.encryptedText.lastIndexOf(')') + 1),
    'base64'
  ).length;
  const bare = encryptDescriptor(descriptorVector.descriptor, { padSecretsTo: 0 }).backup.length;
  const shipped = encryptDescriptor(descriptorVector.descriptor).backup.length;

  // Pinned so a format change is visible, and so the fee it costs a client is
  // a number we state rather than guess.
  //
  // There is no size cap to fear here. Measured on Bitcoin Core v31.1 with
  // default policy (2026-09-10): maxdatacarriersize is 100,000 bytes, and
  // OP_RETURN payloads of 345, 591, 655, 1,000, 10,000, 40,000 and 99,000
  // bytes were all accepted as standard. The only cliff is at 83 bytes, which
  // is where Bitcoin Knots and Core 29 and older stop relaying, and every size
  // here is already past it. So size is a fee question, not a relay question:
  // about 2 sat per extra byte at 2 sat/vB.
  assert.equal(kOfN, 345, 'k-of-n 2-of-3 payload');
  assert.equal(bare, 591, 'BIP-138 2-of-3 with no decoys');
  assert.equal(shipped, 719, 'BIP-138 2-of-3 as Bitcoin Inheritance ships it, padded to seven entries');
});

File: internal/descriptorbackup/bip138_test.go
package descriptorbackup

import (
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"testing"

	"golang.org/x/crypto/chacha20poly1305"
)

const vectorDir = "../html/assets/src/crypto/testdata/bip138"

func loadVectors(t *testing.T, name string, out any) {
	t.Helper()
	b, err := os.ReadFile(filepath.Join(vectorDir, name))
	if err != nil {
		t.Fatalf("reading %s: %v", name, err)
	}
	if err := json.Unmarshal(b, out); err != nil {
		t.Fatalf("parsing %s: %v", name, err)
	}
}

func mustHex(t *testing.T, s string) []byte {
	t.Helper()
	b, err := hex.DecodeString(s)
	if err != nil {
		t.Fatalf("bad hex %q: %v", s, err)
	}
	return b
}

func TestVectorEncryptionSecret(t *testing.T) {
	var vs []struct {
		Description       string   `json:"description"`
		Keys              []string `json:"keys"`
		DecryptionSecret  string   `json:"decryption_secret"`
		IndividualSecrets []string `json:"individual_secrets"`
	}
	loadVectors(t, "encryption_secret.json", &vs)
	if len(vs) == 0 {
		t.Fatal("no vectors")
	}
	for _, v := range vs {
		t.Run(v.Description, func(t *testing.T) {
			keys := make([][]byte, 0, len(v.Keys))
			for _, k := range v.Keys {
				keys = append(keys, mustHex(t, k))
			}
			secret, individual, err := DeriveSecrets(keys)
			if err != nil {
				t.Fatalf("DeriveSecrets: %v", err)
			}
			if got := hex.EncodeToString(secret); got != v.DecryptionSecret {
				t.Errorf("decryption secret\n got %s\nwant %s", got, v.DecryptionSecret)
			}
			if len(individual) != len(v.IndividualSecrets) {
				t.Fatalf("individual secret count: got %d want %d", len(individual), len(v.IndividualSecrets))
			}
			for i, want := range v.IndividualSecrets {
				if got := hex.EncodeToString(individual[i]); got != want {
					t.Errorf("individual secret %d\n got %s\nwant %s", i, got, want)
				}
			}
		})
	}
}

func TestVectorDerivationPath(t *testing.T) {
	var vs []struct {
		Description string   `json:"description"`
		Paths       []string `json:"paths"`
		Expected    *string  `json:"expected"`
	}
	loadVectors(t, "derivation_path.json", &vs)
	for _, v := range vs {
		t.Run(v.Description, func(t *testing.T) {
			paths := make([][]uint32, 0, len(v.Paths))
			for _, p := range v.Paths {
				children, err := ParseDerivationPath(p)
				if err != nil {
					// The draft marks a case that must be refused with a null
					// expectation, and some of those are refused at parse time.
					if v.Expected == nil {
						return
					}
					t.Fatalf("ParseDerivationPath(%q): %v", p, err)
				}
				paths = append(paths, children)
			}
			got, err := EncodeDerivationPaths(paths)
			if v.Expected == nil {
				if err == nil {
					t.Fatalf("expected a refusal, got %s", hex.EncodeToString(got))
				}
				return
			}
			if err != nil {
				t.Fatalf("EncodeDerivationPaths: %v", err)
			}
			if hex.EncodeToString(got) != *v.Expected {
				t.Errorf("\n got %s\nwant %s", hex.EncodeToString(got), *v.Expected)
			}
		})
	}
}

func TestVectorIndividualSecrets(t *testing.T) {
	var vs []struct {
		Description string   `json:"description"`
		Secrets     []string `json:"secrets"`
		Expected    *string  `json:"expected"`
	}
	loadVectors(t, "individual_secrets.json", &vs)
	for _, v := range vs {
		t.Run(v.Description, func(t *testing.T) {
			secrets := make([][]byte, 0, len(v.Secrets))
			for _, s := range v.Secrets {
				secrets = append(secrets, mustHex(t, s))
			}
			got, err := EncodeIndividualSecrets(secrets)
			if v.Expected == nil {
				if err == nil {
					t.Fatalf("expected a refusal, got %s", hex.EncodeToString(got))
				}
				return
			}
			if err != nil {
				t.Fatalf("EncodeIndividualSecrets: %v", err)
			}
			if hex.EncodeToString(got) != *v.Expected {
				t.Errorf("\n got %s\nwant %s", hex.EncodeToString(got), *v.Expected)
			}
		})
	}
}

func TestVectorContentType(t *testing.T) {
	var vs []struct {
		Description string `json:"description"`
		Valid       bool   `json:"valid"`
		Content     string `json:"content"`
	}
	loadVectors(t, "content_type.json", &vs)
	checked := 0
	for _, v := range vs {
		if !v.Valid || len(v.Content) != 6 || v.Content[:2] != "01" {
			continue
		}
		t.Run(v.Description, func(t *testing.T) {
			raw := mustHex(t, v.Content)
			bip := int(raw[1])<<8 | int(raw[2])
			got, err := EncodeBipContentType(bip)
			if err != nil {
				t.Fatalf("EncodeBipContentType: %v", err)
			}
			if hex.EncodeToString(got) != v.Content {
				t.Errorf("\n got %s\nwant %s", hex.EncodeToString(got), v.Content)
			}
		})
		checked++
	}
	if checked == 0 {
		t.Fatal("no BIP-number content-type vectors were exercised")
	}
}

func TestVectorEncryptedBackup(t *testing.T) {
	type extraItem struct {
		Content   string `json:"content"`
		Plaintext string `json:"plaintext"`
	}
	var vs []struct {
		Description            string      `json:"description"`
		Valid                  *bool       `json:"valid"`
		Content                string      `json:"content"`
		Keys                   []string    `json:"keys"`
		DecoyIndividualSecrets []string    `json:"decoy_individual_secrets"`
		DerivationPaths        []string    `json:"derivation_paths"`
		Plaintext              string      `json:"plaintext"`
		Extra                  []extraItem `json:"extra"`
		Nonce                  string      `json:"nonce"`
		Expected               string      `json:"expected"`
	}
	loadVectors(t, "encrypted_backup.json", &vs)

	checked := 0
	for _, v := range vs {
		if v.Valid != nil && !*v.Valid {
			continue
		}
		// The encoder writes BIP-number content items. A vector with a
		// vendor-specific type is covered by the decode test instead.
		if len(v.Content) < 2 || v.Content[:2] != "01" {
			continue
		}
		t.Run(v.Description, func(t *testing.T) {
			keys := make([][]byte, 0, len(v.Keys))
			for _, k := range v.Keys {
				keys = append(keys, mustHex(t, k))
			}
			decoys := make([][]byte, 0, len(v.DecoyIndividualSecrets))
			for _, d := range v.DecoyIndividualSecrets {
				decoys = append(decoys, mustHex(t, d))
			}
			paths := make([][]uint32, 0, len(v.DerivationPaths))
			for _, p := range v.DerivationPaths {
				children, err := ParseDerivationPath(p)
				if err != nil {
					t.Fatalf("ParseDerivationPath(%q): %v", p, err)
				}
				paths = append(paths, children)
			}

			// plaintext is UTF-8 text in these vectors, never hex.
			items := []ContentItem{{Type: ContentTypeBIP, BIP: bipFromContent(t, v.Content), Content: []byte(v.Plaintext)}}
			for _, e := range v.Extra {
				items = append(items, ContentItem{Type: ContentTypeBIP, BIP: bipFromContent(t, e.Content), Content: []byte(e.Plaintext)})
			}

			got, err := EncodeBackup(EncodeOptions{
				Pubkeys:         keys,
				Items:           items,
				DecoySecrets:    decoys,
				DerivationPaths: paths,
				Nonce:           mustHex(t, v.Nonce),
			})
			if err != nil {
				t.Fatalf("EncodeBackup: %v", err)
			}
			if hex.EncodeToString(got) != v.Expected {
				t.Errorf("\n got  %s\nwant  %s", hex.EncodeToString(got), v.Expected)
			}
		})
		checked++
	}
	if checked == 0 {
		t.Fatal("no whole-backup vectors were exercised")
	}
}

// bipFromContent reads the BIP number out of a content-type header hex string
// such as "01017c", the way the draft's vectors express it.
func bipFromContent(t *testing.T, content string) int {
	t.Helper()
	n, err := strconv.ParseInt(content[2:], 16, 32)
	if err != nil {
		t.Fatalf("bad content type %q: %v", content, err)
	}
	return int(n)
}

func TestVectorCipher(t *testing.T) {
	var vs []struct {
		Description string  `json:"description"`
		Nonce       string  `json:"nonce"`
		Plaintext   string  `json:"plaintext"`
		Secret      string  `json:"secret"`
		Ciphertext  *string `json:"ciphertext"`
	}
	loadVectors(t, "chacha20poly1305_encryption.json", &vs)
	checked := 0
	for _, v := range vs {
		if v.Ciphertext == nil {
			continue
		}
		t.Run(v.Description, func(t *testing.T) {
			aead, err := chacha20poly1305.New(mustHex(t, v.Secret))
			if err != nil {
				t.Fatalf("chacha20poly1305.New: %v", err)
			}
			sealed := aead.Seal(nil, mustHex(t, v.Nonce), mustHex(t, v.Plaintext), nil)
			if hex.EncodeToString(sealed) != *v.Ciphertext {
				t.Errorf("\n got %s\nwant %s", hex.EncodeToString(sealed), *v.Ciphertext)
			}
		})
		checked++
	}
	if checked == 0 {
		t.Fatal("no cipher vectors were exercised")
	}
}

// buildBackupWithPayload assembles a container by hand around a payload that
// is already sealed, so a test can put unreadable contents inside a backup the
// key genuinely opens.
func buildBackupWithPayload(t *testing.T, pubkey []byte, payload []byte) []byte {
	t.Helper()
	secret, individual, err := DeriveSecrets([][]byte{pubkey})
	if err != nil {
		t.Fatalf("DeriveSecrets: %v", err)
	}
	nonce := []byte{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12}
	aead, err := chacha20poly1305.New(secret)
	if err != nil {
		t.Fatalf("chacha20poly1305.New: %v", err)
	}
	ciphertext := aead.Seal(nil, nonce, payload, nil)

	secretBlock, err := EncodeIndividualSecrets(individual)
	if err != nil {
		t.Fatalf("EncodeIndividualSecrets: %v", err)
	}
	out := append([]byte{}, Magic...)
	out = append(out, Version)
	out = append(out, 0x00) // no derivation paths
	out = append(out, secretBlock...)
	out = append(out, EncryptionChaCha20Poly1305)
	out = append(out, nonce...)
	out = append(out, EncodeCompactSize(len(ciphertext))...)
	return append(out, ciphertext...)
}

func TestUnreadableContentsDoNotBlameTheKey(t *testing.T) {
	pubkey := mustHex(t, "02e6642fd69bd211f93f7f1f36ca51a26a5290eb2dd1b0d8279a87bb0d480c8443")
	// Content type 0x80 and above is one this version must refuse.
	backup := buildBackupWithPayload(t, pubkey, []byte{0x80, 0x00, 0x00})

	_, err := DecryptBackup(backup, pubkey)
	if err == nil {
		t.Fatal("expected a refusal for contents this version cannot read")
	}
	msg := err.Error()
	if strings.Contains(msg, "none of the keys") {
		t.Errorf("the key was correct, so the error must not blame it:\n  %s", msg)
	}
	if !strings.Contains(msg, "opened this backup") {
		t.Errorf("the error should say the key worked:\n  %s", msg)
	}
}

func TestWrongKeyStillBlamesTheKey(t *testing.T) {
	pubkey := mustHex(t, "02e6642fd69bd211f93f7f1f36ca51a26a5290eb2dd1b0d8279a87bb0d480c8443")
	other := mustHex(t, "0339710356a496726c84692621b2b6e3645dd35bc0026c587f16411897990a1e1f")
	backup := buildBackupWithPayload(t, pubkey, []byte{0x01, 0x01, 0x7c, 0x01, 0x78})

	_, err := DecryptBackup(backup, other)
	if err == nil {
		t.Fatal("a backup must not open with a key that is not in it")
	}
	if !strings.Contains(err.Error(), "none of the keys") {
		t.Errorf("a genuinely wrong key should say so:\n  %s", err.Error())
	}
}
